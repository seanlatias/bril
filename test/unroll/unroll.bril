main {
  i: int = const 0;
  v0: int[3] = new int[3];
  v1: int[3] = new int[3];
  v2: int[3] = new int[3];
  v11: int = const 0;
  v22: int = const 1;
  v33: int = const 2;

  v3: int = const 5;
  set v0 v11 v3;
  v3: int = const 4;
  set v0 v22 v3;
  v3: int = const 8;
  set v0 v33 v3;

  v3: int = const 5;
  set v1 v11 v3;
  v3: int = const 4;
  set v1 v22 v3;
  v3: int = const 8;
  set v1 v33 v3;

for.cond:
  v5: int = id i;
  v6: int = const 3;
  v7: bool = lt v5 v6;
  br v7 for.body for.end;

for.body:
  v8: int = index v0 i;
  v9: int = index v1 i;
  v10: int = add v8 v9;
  set v2 i v10; 
  v12: int = add i v22; 
  i : int = id v12;
  jmp for.cond;

for.end:
  v11: int = index v2 v22;
  print v11;
}
