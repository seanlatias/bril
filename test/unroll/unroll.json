{
  "functions": [
    {
      "instrs": [
        {
          "dest": "i",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "v0",
          "op": "new",
          "type": {
            "base": "int",
            "size": 3
          },
          "type_exp": {
            "base": "int",
            "size": 3
          }
        },
        {
          "dest": "v1",
          "op": "new",
          "type": {
            "base": "int",
            "size": 3
          },
          "type_exp": {
            "base": "int",
            "size": 3
          }
        },
        {
          "dest": "v2",
          "op": "new",
          "type": {
            "base": "int",
            "size": 3
          },
          "type_exp": {
            "base": "int",
            "size": 3
          }
        },
        {
          "dest": "v11",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "v22",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "dest": "v33",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "dest": "v3",
          "op": "const",
          "type": "int",
          "value": 5
        },
        {
          "args": [
            "v0",
            "v11",
            "v3"
          ],
          "op": "set"
        },
        {
          "dest": "v3",
          "op": "const",
          "type": "int",
          "value": 4
        },
        {
          "args": [
            "v0",
            "v22",
            "v3"
          ],
          "op": "set"
        },
        {
          "dest": "v3",
          "op": "const",
          "type": "int",
          "value": 8
        },
        {
          "args": [
            "v0",
            "v33",
            "v3"
          ],
          "op": "set"
        },
        {
          "dest": "v3",
          "op": "const",
          "type": "int",
          "value": 5
        },
        {
          "args": [
            "v1",
            "v11",
            "v3"
          ],
          "op": "set"
        },
        {
          "dest": "v3",
          "op": "const",
          "type": "int",
          "value": 4
        },
        {
          "args": [
            "v1",
            "v22",
            "v3"
          ],
          "op": "set"
        },
        {
          "dest": "v3",
          "op": "const",
          "type": "int",
          "value": 8
        },
        {
          "args": [
            "v1",
            "v33",
            "v3"
          ],
          "op": "set"
        },
        {
          "label": "for.cond"
        },
        {
          "args": [
            "i"
          ],
          "dest": "v5",
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v6",
          "op": "const",
          "type": "int",
          "value": 3
        },
        {
          "args": [
            "v5",
            "v6"
          ],
          "dest": "v7",
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "v7",
            "for.body",
            "for.end"
          ],
          "op": "br"
        },
        {
          "label": "for.body"
        },
        {
          "args": [
            "v0",
            "i"
          ],
          "dest": "v8",
          "op": "index",
          "type": "int"
        },
        {
          "args": [
            "v1",
            "i"
          ],
          "dest": "v9",
          "op": "index",
          "type": "int"
        },
        {
          "args": [
            "v8",
            "v9"
          ],
          "dest": "v10",
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v2",
            "i",
            "v10"
          ],
          "op": "set"
        },
        {
          "args": [
            "i",
            "v22"
          ],
          "dest": "v12",
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v12"
          ],
          "dest": "i",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "for.cond"
          ],
          "op": "jmp"
        },
        {
          "label": "for.end"
        },
        {
          "args": [
            "v2",
            "v22"
          ],
          "dest": "v11",
          "op": "index",
          "type": "int"
        },
        {
          "args": [
            "v11"
          ],
          "op": "print"
        }
      ],
      "name": "main"
    }
  ]
}
