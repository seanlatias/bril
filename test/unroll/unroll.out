#include <stdint.h>
#include <stdio.h>
#include <inttypes.h>
int main(){
int64_t i;
int64_t v11;
int64_t v22;
int64_t v33;
int64_t v3;
int64_t v5;
int64_t v6;
int v7;
int64_t v8;
int64_t v9;
int64_t v10;
int64_t v12;
i = 0;
int v0[3];
int v1[3];
int v2[3];
v11 = 0;
v22 = 1;
v33 = 2;
v3 = 5;
v0[v11] = v3;
v3 = 4;
v0[v22] = v3;
v3 = 8;
v0[v33] = v3;
v3 = 5;
v1[v11] = v3;
v3 = 4;
v1[v22] = v3;
v3 = 8;
v1[v33] = v3;
for.cond:;
v5 = i;
v6 = 3;
v7 = v5 < v6;
if(v7)
    goto for.body;
else
    goto for.end;
for.body:;
v8 = v0[i];
v9 = v1[i];
v10 = v8 + v9;
v2[i] = v10;
v12 = i + v22;
i = v12;
goto for.cond;
for.end:;
v11 = v2[v22];
printf("%" PRId64 "\n", v11);
return 0;
}
